In [1]:
import os
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import feather
import datetime
import itertools
from sklearn import preprocessing
from sklearn.model_selection import LeaveOneGroupOut
from sklearn.model_selection import KFold
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import mean_squared_error
from sklearn.decomposition import TruncatedSVD
from sklearn.decomposition import LatentDirichletAllocation
pd.set_option('display.max_Columns', 100)

In [2]:
train = pd.read_csv('../inputs/train.csv')
print(train.shape)
test = pd.read_csv('../inputs/test.csv')
print(test.shape)

(8359, 16)
(8360, 11)


In [3]:
def creansing(df_input):
    df_input['User_Score_tbd'] = (df_input['User_Score'] == 'tbd').astype(int)
    df_input['User_Score'] = df_input['User_Score'].replace({'nan': np.NaN, 'tbd': np.NaN}).astype(float)
    df_input['User_Score_int'] = df_input['User_Score'].round()
    df_input['Critic_Score_int'] = df_input['Critic_Score'].round(-1)/10
    df_input['User_Count_log'] = df_input['User_Count'].apply(np.log1p)
    df_input['Year'] = df_input['Year_of_Release'].astype(float)
    df_input.drop(columns=['Year_of_Release'], inplace=True)
    return df_input

In [4]:
def fill_null_year(train, test):
    whole = pd.concat([train, test])
    year = whole.groupby(['Publisher', 'Platform'])['Year'].median().reset_index()
    year = year.fillna(0).rename(columns={'Year': 'MedYear'})
    train = pd.merge(train, year, on=['Publisher', 'Platform'], how='left')
    test = pd.merge(test, year, on=['Publisher', 'Platform'], how='left')
    train.loc[train['Year'].isnull(), 'Year'] = train['MedYear']
    test.loc[test['Year'].isnull(), 'Year'] = test['MedYear']
    return train.drop(columns='MedYear'), test.drop(columns='MedYear')

In [5]:
def fill_null_developer(train, test):
    whole = pd.concat([train, test])
    
    # 'Genre', 'Platform'で最大数の'Developer'で埋める
    mode_dev = whole.groupby(['Genre', 'Platform'])[['Developer']].apply(lambda x: x.mode()).reset_index()
    mode_dev = mode_dev.drop_duplicates(subset=['Genre', 'Platform'])
    mode_dev = mode_dev.rename(columns={'Developer':'devlop'})
    train = pd.merge(train, mode_dev, on=['Genre', 'Platform'], how='left')
    test = pd.merge(test, mode_dev, on=['Genre', 'Platform'], how='left')
    train.loc[train['Developer'].isnull(), 'Developer'] = train['devlop']
    test.loc[test['Developer'].isnull(), 'Developer'] = test['devlop']

    return train.drop(columns='devlop'), test.drop(columns='devlop')

In [6]:
train = creansing(train)
test = creansing(test)
train, test = fill_null_year(train, test)
train, test = fill_null_developer(train, test)

In [7]:
n_fold = 5

### label_encode

In [8]:
def get_non_overlapping(train: pd.DataFrame, test: pd.DataFrame, column: str):
    """train/testにしか出てこない値を調べる"""
    only_in_train = set(train[column].unique()) - set(test[column].unique())
    only_in_test = set(test[column].unique()) - set(train[column].unique())
    non_overlapping = only_in_train.union(only_in_test)
    return non_overlapping

def category2num(train: pd.DataFrame, test: pd.DataFrame, columns: list):
    train_ = train.copy()
    test_ = test.copy()
    for column in columns:
        non_overlapping = get_non_overlapping(train, test, column)
        try:
            if train[column].dtype == np.dtype("O"):
                # dtypeがobjectなら欠損は'missing' クラスにする
                train_[column] = train[column].fillna("missing")
                test_[column] = test[column].fillna("missing")
                train_[column] = train_[column].map(lambda x: x if x not in non_overlapping else "other")
                test_[column] = test_[column].map(lambda x: x if x not in non_overlapping else "other")
            else:
                # dtypeがint/floatなら欠損は'-1'とする
                train_[column] = train[column].fillna(-1)
                test_[column] = test[column].fillna(-1)
                train_[column] = train_[column].map(lambda x: x if x not in non_overlapping else -2)
                test_[column] = test_[column].map(lambda x: x if x not in non_overlapping else -2)

            le = preprocessing.LabelEncoder()
            concatenated = pd.concat([train_, test_], axis=0).reset_index(drop=True)
            le.fit(concatenated[column])
            train_[column] = le.transform(train_[column])
            test_[column] = le.transform(test_[column])
        except Exception:
            import pdb
            pdb.set_trace()
    return train_, test_

In [9]:
def create_label_encoding(df_train, df_test):
    tr, te = category2num(df_train, df_test, ['Platform', 'Genre', 'Developer', 'Rating'])
    return tr, te

### count_encode

In [10]:
def count_encoding(df_train, df_test, col):
    tr = df_train[col]
    te = df_test[col]
    whole = pd.concat([tr, te])
    vc = whole.value_counts(dropna=False)
    tr_ = pd.DataFrame(tr.map(vc)).add_prefix('CE_')
    te_ = pd.DataFrame(te.map(vc)).add_prefix('CE_')
    return tr_, te_

def create_count_encoding_feature(df_train, df_test):
    tr = pd.DataFrame()
    te = pd.DataFrame()
    for col in ['Platform', 'Genre', 'Publisher', 'Developer', 'Rating', 'User_Score_int', 'Critic_Score_int']:
        tr_, te_ = count_encoding(df_train, df_test, col)
        tr = pd.concat([tr, tr_], axis=1)
        te = pd.concat([te, te_], axis=1)
    return tr, te

### target_encode

In [11]:
def target_encoding(src, dst, group_col, target_col, aggfunc):
    agg_dict = {}
    for tc in target_col:
        agg_dict[tc] = aggfunc
    target = src.groupby(group_col).agg(agg_dict).reset_index()
    
    merge = pd.merge(dst[group_col], target, on=group_col, how='left').set_index(dst.index)
    suffix = '_'.join(group_col)
    return merge.drop(columns=group_col).add_prefix('TGE_').add_suffix('_{}_by_{}'.format(aggfunc, suffix))

In [12]:
def crete_target_enc_train_test(df_train, df_test, group_col, target_col, aggfunc):
    target = pd.DataFrame()
    kf = KFold(n_splits=n_fold, shuffle=True, random_state=0)
    for train_index, eval_index in kf.split(df_train):
        kf_train = df_train.iloc[train_index]
        kf_eval = df_train.iloc[eval_index]
        tmp = target_encoding(kf_train, kf_eval, group_col, target_col, aggfunc)
        target = pd.concat([target, tmp])
    # train
    tr = target.sort_index()
    # test
    te = target_encoding(df_train, df_test, group_col, target_col, aggfunc)
    return tr, te

In [13]:
def create_target_encoding_sub(df_train, df_test, tr, te, group, target, aggfunc):
    tr_, te_ = crete_target_enc_train_test(df_train, df_test, group, target, aggfunc)
    tr = pd.concat([tr, tr_], axis=1)
    te = pd.concat([te, te_], axis=1)
    return tr, te

In [14]:
def create_target_encoding(df_train, df_test):
    tr = pd.DataFrame()
    te = pd.DataFrame()
#     target = ['Global_Sales', 'NA_Sales', 'EU_Sales', 'JP_Sales', 'Other_Sales']
    target = ['Global_Sales']
    groups = [
        ['Platform'],
        ['Genre'],
        ['Rating'],
        ['User_Score_int'],
        ['Critic_Score_int'],
        
        ['Platform', 'Genre'],
        ['Platform', 'Rating'],
        ['Genre', 'Rating'],
        
        ['User_Score_int', 'Platform'],
        ['User_Score_int', 'Genre'],
        ['User_Score_int', 'Rating'],
        ['Critic_Score_int', 'Platform'],
        ['Critic_Score_int', 'Genre'],
        ['Critic_Score_int', 'Rating'],
        ['Critic_Score_int', 'User_Score_int'],
    ]
    for gr in groups:
        tr, te = create_target_encoding_sub(df_train, df_test, tr, te, gr, target, 'mean')
    for gr in groups:
        tr, te = create_target_encoding_sub(df_train, df_test, tr, te, gr, target, 'std')
    
    return tr, te

### word (TruncatedSVD)

In [15]:
def word_pivot(df_input, wordlist):
    pivot = pd.DataFrame(index=df_input.index, columns=wordlist)
    pivot = pd.concat([df_input['Name'].str.lower(), pivot], axis=1)
    for c in pivot.columns:
        if c == 'Name':
            continue
        pivot[c] = pivot['Name'].str.contains(c)
    return pivot

In [16]:
def create_wordlist():
    trainName = train['Name'].str.lower()
    testName = test['Name'].str.lower()
    name = pd.concat([trainName, testName]).reset_index(drop=True)
    clean_name = name.str.replace('[^0-9A-Za-z]', ' ')
    
    wordlist = []
    for n in list(clean_name):
        wordlist = wordlist + (str(n).split())
    print(len(wordlist))
    
    df_name = pd.DataFrame(pd.DataFrame(wordlist, columns=['name']).value_counts()).reset_index()
    df_name.rename(columns={0:'name_count'}, inplace=True)
    
    # 3文字以上、3回以上出現している単語
    df_name['word_len'] = df_name['name'].str.len()
    df_longname = df_name[df_name['word_len'] > 3]
    df_longname = df_longname[df_longname['name_count'] > 3]
    
    # 含まれるかどうか
    train_pivot = word_pivot(train, list(df_longname['name']))
    print('train_pivot', train_pivot.shape)
    test_pivot = word_pivot(test, list(df_longname['name']))
    print('test_pivot', test_pivot.shape)
    
    # trainとtestの両方に存在するものだけ(train>3, test>0)
    train_sum = pd.DataFrame(train_pivot.drop(columns=['Name']).sum(), columns=['tr_count'])
    test_sum = pd.DataFrame(test_pivot.drop(columns=['Name']).sum(), columns=['te_count'])
    train_test_sum = pd.concat([train_sum, test_sum], axis=1).reset_index()
    new_wordlist = train_test_sum[(train_test_sum['tr_count']>3)&(train_test_sum['te_count']>0)].reset_index(drop=True)
    
    train_pivot2 = word_pivot(train, list(new_wordlist['index']))
    print('train_pivot2', train_pivot2.shape)
    test_pivot2 = word_pivot(test, list(new_wordlist['index']))
    print('test_pivot2', test_pivot2.shape)
    
    train_pivot2.to_feather('../inputs/train_word.f')
    test_pivot2.to_feather('../inputs/test_word.f')
    
    return train_pivot2, test_pivot2

In [17]:
skip_create_wordlist = True

In [18]:
def create_word_features(df_train, df_test):
    
    if skip_create_wordlist:
        train_word = pd.read_feather('../inputs/train_word.f')
        test_word = pd.read_feather('../inputs/test_word.f')
    else:
        train_word, test_word = create_wordlist()
        
    wd_tr = train_word.drop(columns=['Name']).fillna(0).astype(int).add_prefix('WD_')
    wd_te = test_word.drop(columns=['Name']).fillna(0).astype(int).add_prefix('WD_')
    print('wd_tr', wd_tr.shape)
    print('wd_te', wd_te.shape)
    
    return wd_tr, wd_te

#     ntopic = 5
#     lda = LatentDirichletAllocation(n_components=ntopic)
#     x_tr = pd.DataFrame(lda.fit_transform(wd_tr)).add_prefix('LDA_')
#     x_te = pd.DataFrame(lda.fit_transform(wd_te)).add_prefix('LDA_')
    
#     svd = TruncatedSVD(5)
#     x_tr = pd.DataFrame(svd.fit_transform(wd_tr)).add_prefix('SVD_')
#     x_te = pd.DataFrame(svd.fit_transform(wd_te)).add_prefix('SVD_')
#     return x_tr, x_te
    
#     tr_ = pd.concat([wd_tr, x_tr], axis=1)
#     te_ = pd.concat([wd_te, x_te], axis=1)
#     return tr_, te_

### LDA

In [19]:
def LDA_topic(df_input, ntopic, index, column):

    df_input['dummy'] = 1
    pvt_tbl = pd.pivot_table(df_input, index=index, columns=column, values='dummy', aggfunc='count')
    pvt_tbl.fillna(0, inplace=True)
    
    prefix = 'LDA_{}_'.format(column)
    suffix = '_by_{}'.format(index)
    
    lda = LatentDirichletAllocation(n_components=ntopic)
    lda_out = pd.DataFrame(lda.fit_transform(pvt_tbl), index=pvt_tbl.index).add_prefix(prefix).add_suffix(suffix)
    return lda_out.reset_index()

In [20]:
def create_lda_features(df_train, df_test):
    cat_list = ['Platform', 'Genre', 'Developer', 'Rating', 'Publisher']
    tr = df_train[cat_list].copy()
    te = df_test[cat_list].copy()
    whole = pd.concat([tr, te])
    
    for a, b in itertools.permutations(cat_list, 2):
        lda_o = LDA_topic(whole, 5, index=a, column=b)
        tr = pd.merge(tr, lda_o, on=a, how='left')
        te = pd.merge(te, lda_o, on=a, how='left')
        
    return tr.drop(columns=cat_list), te.drop(columns=cat_list)

### salesの割合

In [21]:
sales_columns = ['NA_Sales', 'EU_Sales', 'JP_Sales', 'Other_Sales'] #, 'Global_Sales']

def sales_groupby(df_train, df_test, groupkey):
    tr = df_train[groupkey].copy()
    te = df_test[groupkey].copy()
    grp_sum = df_train[sales_columns].groupby(df_train[groupkey]).sum()
    
    # 列方向の割合
#     salescol_sum = grp_sum.sum()
#     for c in sales_columns:
#         grp_sum['Rate1_'+c] = grp_sum[c] / salescol_sum[c]
#     grp_sum.drop(columns=['Rate1_Global_Sales'], inplace=True)
    
    for c in sales_columns:
        grp_sum['Rank_'+c] = grp_sum[c].rank()

    grp_sum.drop(columns=sales_columns, inplace=True)
    grp_sum = grp_sum.add_suffix('_by_{}'.format(groupkey))

    tr = pd.merge(tr, grp_sum.reset_index(), on=groupkey, how='left').drop(columns=groupkey)
    te = pd.merge(te, grp_sum.reset_index(), on=groupkey, how='left').drop(columns=groupkey)
    return tr, te

In [22]:
def create_sales_feature(df_train, df_test):
    tr = pd.DataFrame(index=df_train.index)
    te = pd.DataFrame(index=df_test.index)
    keys = ['Platform', 'Genre', 'Developer', 'Rating', 'Year']
    for k in keys:
        tr_, te_ = sales_groupby(train, test, k)
        tr = pd.concat([tr, tr_], axis=1)
        te = pd.concat([te, te_], axis=1)
    return tr, te

### publisher

In [23]:
def create_publisher_feature(df_train, df_test):
    df_whole = pd.concat([df_train, df_test])
    plat_pivot = df_whole.pivot_table(index='Publisher', columns='Platform', values='Name', aggfunc='count')
    genre_pivot = df_whole.pivot_table(index='Publisher', columns='Genre', values='Name', aggfunc='count')
    year_pivot = df_whole.pivot_table(index='Publisher', columns='Year', values='Name', aggfunc='count')
    print('plat_pivot', len(plat_pivot))
    print('genre_pivot', len(genre_pivot))
    print('year_pivot', len(year_pivot))
    
    all_pivot = pd.concat([plat_pivot, genre_pivot], axis=1)
    all_pivot = pd.concat([all_pivot, year_pivot], axis=1)
    all_pivot.fillna(0, inplace=True)
    print('all_pivot', len(all_pivot))
    
    lda = LatentDirichletAllocation(n_components=10)
    lda_out = pd.DataFrame(lda.fit_transform(all_pivot), index=all_pivot.index).add_prefix('LDA_Publisher_').reset_index()
    
    tr = pd.merge(df_train[['Publisher']], lda_out, on='Publisher', how='left').drop(columns=['Publisher'])
    te = pd.merge(df_test[['Publisher']], lda_out, on='Publisher', how='left').drop(columns=['Publisher'])
    print(tr.shape)
    print(te.shape)
    return tr, te

### 特徴量生成

In [24]:
processors = [
#     create_sales_feature,
    create_count_encoding_feature,
    create_target_encoding,
    create_lda_features,
    create_word_features,
    create_publisher_feature,
]

In [25]:
def create_feature(train, test):
    train_cp = train.copy()
    test_cp = test.copy()
    
    for func in processors:
        tr, te = func(train_cp, test_cp)
        train_cp = pd.concat([train_cp, tr], axis=1)
        test_cp = pd.concat([test_cp, te], axis=1)

    # label enc
#     train_cp, test_cp = create_label_encoding(train_cp, test_cp)
    
    print('train', train_cp.shape)
    print('test', test_cp.shape)
    assert len(train) == len(train_cp)
    assert len(test) == len(test_cp)
    return train_cp, test_cp

In [26]:
train_, test_ = create_feature(train, test)

wd_tr (8359, 1271)
wd_te (8360, 1271)
plat_pivot 581
genre_pivot 581
year_pivot 581
all_pivot 581
(8359, 10)
(8360, 10)
train (8359, 1439)
test (8360, 1434)


In [27]:
train_.to_feather('../inputs/train_cat_feature.f')
test_.to_feather('../inputs/test_cat_feature.f')

In [28]:
# label enc
train_, test_ = create_label_encoding(train_, test_)
train_.to_feather('../inputs/train_feature.f')
test_.to_feather('../inputs/test_feature.f')

assert len(train) == len(train_)
assert len(test) == len(test_)

#### 25
- Year, DeveloperのNaNを埋める
- train (8359, 1439)
- test (8360, 1434)

#### 23
- SalesのRank
- train (8359, 1459)
- test (8360, 1454)

#### 22
- PublisherごとのcountをLDA


In [29]:
train_.head()

Name  Platform  Genre  \
0                   LEGO Batman: The Videogame        23      0   
1  LEGO Indiana Jones: The Original Adventures        23      0   
2                   LEGO Batman: The Videogame        17      0   
3                                       Combat         0      0   
4                 LEGO Harry Potter: Years 5-7        23      0   

                                Publisher  NA_Sales  EU_Sales  JP_Sales  \
0  Warner Bros. Interactive Entertainment       180        97         0   
1                               LucasArts       151        61         0   
2  Warner Bros. Interactive Entertainment        56        44         0   
3                                   Atari       117         7         0   
4  Warner Bros. Interactive Entertainment        69        42         0   

   Other_Sales  Global_Sales  Critic_Score  Critic_Count  User_Score  \
0           28           306          74.0          17.0         7.9   
1           21           234          78.0          22.0         6.6   
2           27           128          73.0           5.0         7.4   
3            1           125           NaN           NaN         NaN   
4           12           124          76.0           8.0         7.8   

   User_Count  Developer  Rating  User_Score_tbd  User_Score_int  \
0        22.0        454       1               0             8.0   
1        28.0        454       1               0             7.0   
2        10.0        454       1               0             7.0   
3         NaN        497       6               0             NaN   
4        13.0        454       1               0             8.0   

   Critic_Score_int  User_Count_log    Year  level_2  CE_Platform  CE_Genre  \
0               7.0        3.135494  2010.0      0.0         1320      3370   
1               8.0        3.367296  2008.0      0.0         1320      3370   
2               7.0        2.397895  2010.0      0.0         1209      3370   
3               NaN             NaN  1982.0      NaN          133      3370   
4               8.0        2.639057  2010.0      0.0         1320      3370   

   CE_Publisher  CE_Developer  CE_Rating  CE_User_Score_int  \
0           235           237       1420               2830   
1            90           237       1420               1634   
2           235           237       1420               1634   
3           367          1127       6769               9129   
4           235           237       1420               2830   

   CE_Critic_Score_int  TGE_Global_Sales_mean_by_Platform  \
0                 2019                          93.604962   
1                 2366                          96.597250   
2                 2019                          29.606403   
3                 8582                          65.943662   
4                 2366                          96.597250   

   TGE_Global_Sales_mean_by_Genre  TGE_Global_Sales_mean_by_Rating  \
0                       61.883755                        49.257785   
1                       58.394680                        49.148789   
2                       59.681720                        53.568595   
3                       62.435252                              NaN   
4                       58.394680                        49.148789   

   TGE_Global_Sales_mean_by_User_Score_int  \
0                               106.536082   
1                                66.403374   
2                                64.981846   
3                                      NaN   
4                                99.656627   

   TGE_Global_Sales_mean_by_Critic_Score_int  \
0                                  49.231678   
1                                 101.466970   
2                                  46.931655   
3                                        NaN   
4                                 101.466970   

   TGE_Global_Sales_mean_by_Platform_Genre  \
0                                59.678899   
1                                54.090000   
